### ISSUE TITLE
`with torch.enable_grad` also works outside a `no_grad` context

### ISSUE THREAD

**Issue by HaleTom** [pytorch/pytorch#19189]
## 📚 Documentation

The [`torch.autograd.enable_grad` documentation](https://pytorch.org/docs/stable/autograd.html#torch.autograd.enable_grad) says:

> Enables gradient calculation inside a `no_grad` context. This has no effect outside of `no_grad`.

This implies:
```
    torch.set_grad_enabled(False)
    with torch.enable_grad:
        # Gradient tracking will NOT be enabled here.
    torch.set_grad_enabled(True)

vs:

    with torch.no_grad():
        with torch.enable_grad:
            # Gradient tracking IS enabled here.
```
 
However the observed behaviour (`1.0.1.post2`) is:

```
x = torch.tensor([.1], requires_grad=True)

with torch.no_grad():
    with torch.enable_grad():
        y = x * 2
print(y.requires_grad) # True (as expected)

with torch.set_grad_enabled(False):
    y = x * 2
print(y.requires_grad)  # False (as expected)

with torch.set_grad_enabled(False):
    with torch.enable_grad():
        y = x * 2
        print(y.requires_grad)  # True, but False expected from doc quote
```

Note the last example is not "inside a `no_grad` context", but it still works.

Other prior art: [PyTorch set_grad_enabled(False) vs with no_grad():](https://stackoverflow.com/a/53447634/5353461)

-----------

I'm assuming the documentation is incorrect, and it should say simply:

> Enables gradient calculation.

Would you accept a PR?

**Comment by zou3519** [comment]
"Enables gradient calculation, if it has been disabled via `no_grad` or `torch.set_grad_enabled`" <- I think something like that would make it clearer. But yes, we would accept a PR, thank you!

**Comment by HaleTom** [comment]
There is also this case:

```
x = torch.tensor([.1], requires_grad=True)

def doubler(x): return x * 2
y = doubler(x)
print(y.requires_grad) # True (obvious)

# Again, with decorator
@torch.no_grad()
def doubler(x): return x * 2
y = doubler(x)
print(y.requires_grad) # False (obvious)

with torch.enable_grad():
    y = doubler(x)
print(y.requires_grad) # False (non-obvious, subsequently disabled by `doubler()` )
```

@zou3519 Perhaps it's better to look at this from the other perspective:

What are the situations in which `torch.autograd.enable_grad` **doesn't** enable gradients?

How about:

> Enables gradient calculations.
> Note:  gradient calculations may be disabled if the following code calls `no_grad()` or `set_grad_enabled(False)`.

**Comment by HaleTom** [comment]
`no_grad()` and `set_grad_enabled()` should probably have the reciprocal `Note:` for completeness.

**Comment by HaleTom** [comment]
Thanks a bunch! Sorry I didn't get around to it myself!
